[lldb-dap] Ensure the IO forwarding threads are managed by the DAP object lifecycle.

ashgti opened the pull request:
This moves the ownership of the threads that forward stdout/stderr to the DAP object itself to ensure that the threads are joined and that the forwarding is cleaned up when the DAP connection is disconnected.

This is part of a larger refactor to allow lldb-dap to run in a listening mode and accept multiple connections.

ashgti committed:
[lldb-dap] Ensure the IO forwarding threads are managed by the DAP object lifecycle.

This moves the ownership of the threads that forward stdout/stderr to the DAP object itself to ensure that the threads are joined and that the forwarding is cleaned up when the DAP connection is disconnected.

This is part of a larger refactor to allow lldb-dap to run in a listening mode and accept multiple connections.

ashgti committed:
Applying formatting.

ashgti committed:
Applying formatting to lldb-dap.cpp.

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -173,6 +178,63 @@ ExceptionBreakpoint *DAP::GetExceptionBreakpoint(const lldb::break_id_t bp_id) {
+    size_t bytes_read;
+    while (pipe.CanRead()) {
+      lldb_private::Status error = pipe.ReadWithTimeout(
+          &buffer, sizeof(buffer), std::chrono::seconds(1), bytes_read);
```
If this really can be a read-until-EOF loop, then you could just make this a blocking read (no timeout). *However*, I think this is not a good idea to do in combination with duping the pipe to stdout/err (TIL that `dup2` is a thing on windows), because in order to get an EOF you have to close *all* copies of that FD. It looks like you don't do that, which is probably the cause of the presubmit test failures, but even if you did (perhaps by restoring the original FD, or /dev/null), I still wouldn't really like this approach because it's very easy for stdout/err descriptors to be leaked into other processes (which might end up outliving all lldb-dap connections).

labath replied:
How about this: use a blocking read (no timeout) in combination with an `atomic_flag` for termination. When you want to terminate the thread, set the flag and *then* write something to the pipe (to ensure the thread wakes up and sees the flag)?

ashgti replied:
Updated per this suggestion and I moved this logic into a `OutputRedirection` helper in the `OutputRedirection.{h,cpp}` file. I also included a comment to help explain why we need to have a write to ensure the thread stops.

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -173,6 +178,63 @@ ExceptionBreakpoint *DAP::GetExceptionBreakpoint(const lldb::break_id_t bp_id) {
+    }
+  }
+
+  auto forwarder = [&](lldb_private::Pipe &pipe, OutputType outputType) {
```
don't use `[&]` in a lambda which outlives the enclosing function. List the captures explicitly.

ashgti replied:
Done.

labath commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1099,6 +1098,14 @@ void request_disconnect(DAP &dap, const llvm::json::Object &request) {
     dap.progress_event_thread.join();
   }
+  if (dap.stdout_forward_thread.joinable()) {
+    dap.pout.Close();
```
This may be moot due to due other issue, but you should only close the write end before joining the thread (the read end can be closed after joining, or from within the thread). Closing the read descriptor while the thread is running leads to a race with the `Read` call inside the thread.

ashgti replied:
Done.

labath commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5027,43 +5019,77 @@ int main(int argc, char *argv[]) {
+    result = _setmode(fileno(stdin), _O_BINARY);
+    UNUSED_IF_ASSERT_DISABLED(result);
+    assert(result);
+#endif
+
+    int stdout_fd = dup(fileno(stdout));
```
```suggestion
    int stdout_fd = dup(fileno(stdout));
#else
    int stdout_fd = fcntl(fileno(stdout), F_DUPFD_CLOEXEC, 3);
#endif

```

I know this isn't your code, but I think it's a good idea to set this flag, particularly as we're going to be doing (a lot) more things within the same process.

ashgti replied:
I created a help for duplicating the FD and added a `#ifdef` around this like I found in a few other places in lldb like https://github.com/llvm/llvm-project/blob/81831ef3e7804b9637aff119c1adde661d10fb78/lldb/source/Host/posix/PipePosix.cpp#L43-L50

labath replied:
I would work (since the code is still single-threaded at this point), but I don't think it's better than F_DUPFD_CLOEXEC. The reason the code you quote is written this way is because non-linux systems don't have the equivalent of the `pipe2`. I've checked that macos and all supported BSD variants do have this option. The only one I couldn't check is AIX, but it's still too early to call that "supported".

labath commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5027,43 +5019,77 @@ int main(int argc, char *argv[]) {
+    output = StreamDescriptor::from_file(stdout_fd, false);
   }
 
+  DAP dap = DAP(program_path.str(), log, default_repl_mode, std::move(input),
```
If I read this correctly, to get rid of all the references-to-optional, all you'd need is to change this to `log?&*log:nullptr`

ashgti replied:
Done.

ashgti committed:
Moving the output redirection logic back into the OutputRedirector.cpp file and making it into an object instead.

This removes some of the duplicate code that was happening between stdout/stderr redirection.

ashgti commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -173,6 +178,63 @@ ExceptionBreakpoint *DAP::GetExceptionBreakpoint(const lldb::break_id_t bp_id) {
   return nullptr;
 }
 
+llvm::Error DAP::ConfigureIO(std::optional<std::FILE *> overrideOut,
```
Removed the optional and just passed a `std::FILE *` instead.

ashgti commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -138,15 +140,20 @@ struct SendEventRequestHandler : public lldb::SBCommandPluginInterface {
 
 struct DAP {
   llvm::StringRef debug_adaptor_path;
+  std::optional<std::ofstream> &log;
```
Removed the optional and just passed a plain pointer instead.

ashgti commented on lldb/tools/lldb-dap/IOStream.h:
```diff
@@ -52,16 +52,24 @@ struct StreamDescriptor {
+      : descriptor(std::move(descriptor)) {};
 
-  bool read_line(std::ofstream *log, std::string &line);
+  bool read_full(std::optional<std::ofstream> &log, size_t length,
```
Reverted these changes.

ashgti committed:
Simplifying how the log object is created, using a unique_ptr instead of an optional.

labath commented on lldb/tools/lldb-dap/OutputRedirector.h:
```diff
@@ -9,17 +9,38 @@
-///     \a Error::success if the redirection was set up correctly, or an error
-///     otherwise.
-llvm::Error RedirectFd(int fd, std::function<void(llvm::StringRef)> callback);
+struct OutputRedirector {
```
```suggestion
class OutputRedirector {
public:
```

ashgti replied:
Done

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -173,6 +179,55 @@ ExceptionBreakpoint *DAP::GetExceptionBreakpoint(const lldb::break_id_t bp_id) {
+    if (dup2(*fd, fileno(overrideOut)) == -1)
+      return llvm::make_error<llvm::StringError>(
+          llvm::errnoAsErrorCode(),
+          llvm::formatv("override fd=%d failed", fileno(overrideOut))
```
```suggestion
          llvm::formatv("override fd={0} failed", fileno(overrideOut))
```

`formatv` takes python-like replacement strings, but honestly, I'd probably just write this as `return errorCodeToError(llvm::errnoAsErrorCode())`

ashgti replied:
Done

labath commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -138,15 +140,20 @@ struct SendEventRequestHandler : public lldb::SBCommandPluginInterface {
   std::thread event_thread;
   std::thread progress_event_thread;
-  std::unique_ptr<std::ofstream> log;
+  std::thread stdout_forward_thread;
+  std::thread stderr_forward_thread;
```
not used?

ashgti replied:
Removed

labath commented on lldb/tools/lldb-dap/IOStream.h:
```diff
@@ -52,6 +52,9 @@ struct StreamDescriptor {
   StreamDescriptor descriptor;
 
+  explicit InputStream(StreamDescriptor descriptor)
+      : descriptor(std::move(descriptor)) {};
```
```suggestion
      : descriptor(std::move(descriptor)) {}
```

ashgti replied:
Done

labath commented on lldb/tools/lldb-dap/IOStream.h:
```diff
@@ -62,6 +65,9 @@ struct InputStream {
   StreamDescriptor descriptor;
 
+  explicit OutputStream(StreamDescriptor descriptor)
+      : descriptor(std::move(descriptor)) {};
```
```suggestion
      : descriptor(std::move(descriptor)) {}
```

ashgti replied:
Done.

labath commented on lldb/tools/lldb-dap/OutputRedirector.cpp:
```diff
@@ -17,47 +19,59 @@
-      if (bytes_count == -1) {
-        if (errno == EAGAIN || errno == EINTR)
-          continue;
+    while (m_pipe.CanRead() && !m_stopped) {
+      size_t bytes_read;
+      Status status = m_pipe.Read(&buffer, sizeof(buffer), bytes_read);
+      if (status.Fail())
+        continue;
+
+      // EOF detected
+      if (bytes_read == 0)
```
```suggestion
    while (m_pipe.CanRead()) {
      size_t bytes_read;
      Status status = m_pipe.Read(&buffer, sizeof(buffer), bytes_read);
      if (status.Fail())
        continue;

      // EOF detected
      if (bytes_read == 0 || m_stopped)
```

.. motivation being to avoid sending the empty null byte. It's true that this can prevent us from sending other data as well, but I don't think anything can rely on that, as the thread can still exit before it manages to read that data from the pipe. If anything wanted to rely on some output being sent, it would have to implement to form of out-of-band synchronization to make sure the data is received before shutting things down.

ashgti replied:
Done.

labath commented on lldb/tools/lldb-dap/OutputRedirector.cpp:
```diff
@@ -17,47 +19,59 @@
+  m_stopped = true;
+
+  if (m_pipe.CanWrite()) {
+    // If the fd is waiting for input and is closed it may not return from the
+    // current select/poll/kqueue/etc. asyncio wait operation. Write a null byte
+    // to ensure the read fd wakes to detect the closed FD.
```
```suggestion
    // Closing the pipe may not be sufficient to wake up the thread in case the write descriptor is duplicated (to stdout/err or to another process). Write a null byte
    // to ensure the read call returns.
```

(I found the comment confusing -- I don't think it's correct to say that an *fd* is waiting for something. Threads are waiting (on/for the fd).

ashgti replied:
Done

labath commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5030,47 +5045,81 @@ int main(int argc, char *argv[]) {
+  std::unique_ptr<std::ofstream> log = nullptr;
+  const char *log_file_path = getenv("LLDBDAP_LOG");
+  if (log_file_path)
+    log.reset(new std::ofstream(log_file_path));
```
```suggestion
    log = std::make_unique<std::ofstream>(log_file_path));
```

to avoid the bare `new`.

ashgti replied:
Done.

labath commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -5030,47 +5045,81 @@ int main(int argc, char *argv[]) {
+
+  // stdout/stderr redirection to the IDE's console
+  if (auto Err = dap.ConfigureIO(redirectOut, redirectErr)) {
+    llvm::errs() << "Failed to configure lldb-dap IO operations: " << Err
```
```suggestion
    logAllUnhandledErrors(std::move(Err), llvm::errs(), "Failed to configure lldb-dap IO operations: ");
```

Simply streaming the error object will not set the "checked" bit and will cause at assert (if they are enabled).

ashgti replied:
Done.

labath reviewed (commented):
We're pretty close, but I still see some presubmit failures in TestDAP_redirection_to_console, which I have to assume are related to this change.

ashgti committed:
Applying reviewer comments.

ashgti commented:
I'm running into an issue on linux with this change that is resulting in 2 instances of the `lldb_private::FileSystem::Instance` being created. 

```
$ lldb lldb-dap
(lldb) break set -f FileSystem.h -l 51
...

Process 1743394 resuming
/usr/bin/../lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/optional:475: _Tp &std::_Optional_base_impl<lldb_private::FileSystem, std::_Optional_base<lldb_private::FileSystem, false, false>>::_M_get() [_Tp = lldb_private::FileSystem, _Dp = std::_Optional_base<lldb_private::FileSystem, false, false>]: Assertion 'this->_M_is_engaged()' failed.
Process 1743394 stopped
* thread #1, name = 'lldb-dap', stop reason = signal SIGABRT
    frame #0: 0x00007fffec0a53ac libc.so.6`__pthread_kill_implementation(threadid=<unavailable>, signo=6, no_tid=<unavailable>) at pthread_kill.c:44:76
(lldb) bt
* thread #1, name = 'lldb-dap', stop reason = signal SIGABRT
  * frame #0: 0x00007fffec0a53ac libc.so.6`__pthread_kill_implementation(threadid=<unavailable>, signo=6, no_tid=<unavailable>) at pthread_kill.c:44:76
    frame #1: 0x00007fffec0564f2 libc.so.6`__GI_raise(sig=6) at raise.c:26:13
    frame #2: 0x00007fffec03f4ed libc.so.6`__GI_abort at abort.c:79:7
    frame #3: 0x00007fffec2d501e libstdc++.so.6`std::__glibcxx_assert_fail(char const*, int, char const*, char const*) + 94
    frame #4: 0x00005555555d255d lldb-dap`lldb_private::FileSystem::Instance() [inlined] std::_Optional_base_impl<lldb_private::FileSystem, std::_Optional_base<lldb_private::FileSystem, false, false>>::_M_get(this=<unavailable>) at optional:475:2
    frame #5: 0x00005555555d253e lldb-dap`lldb_private::FileSystem::Instance() [inlined] std::optional<lldb_private::FileSystem>::operator*(this=<unavailable>) & at optional:972:22
    frame #6: 0x00005555555d253e lldb-dap`lldb_private::FileSystem::Instance() at FileSystem.cpp:45:45
    frame #7: 0x000055555558fd7f lldb-dap`lldb_dap::DAP::ConfigureIO(this=<unavailable>, overrideOut=<unavailable>, overrideErr=<unavailable>) at DAP.cpp:183:17
    frame #8: 0x000055555556d479 lldb-dap`main(argc=<unavailable>, argv=<unavailable>) at lldb-dap.cpp:5109:22
    frame #9: 0x00007fffec040c8a libc.so.6`__libc_start_call_main(main=(lldb-dap`main at lldb-dap.cpp:4960), argc=1, argv=0x00007fffffffda98) at libc_start_call_main.h:58:16
    frame #10: 0x00007fffec040d45 libc.so.6`__libc_start_main_impl(main=(lldb-dap`main at lldb-dap.cpp:4960), argc=1, argv=0x00007fffffffda98, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=0x00007fffffffda88) at libc-start.c:360:3
    frame #11: 0x000055555556c5f1 lldb-dap`_start + 33

(lldb) break list
Current breakpoints:
1: file = 'FileSystem.h', line = 51, exact_match = 0, locations = 3, resolved = 3, hit count = 2
  1.1: where = liblldb.so.20.0git`lldb_private::SystemInitializerCommon::Initialize() + 23 [inlined] void lldb_private::FileSystem::Initialize<>() at FileSystem.h:51:5, address = 0x00007fffed8c6be7, resolved, hit count = 1 
  1.2: where = liblldb.so.20.0git`lldb_private::SystemInitializerCommon::Initialize() + 32 [inlined] void lldb_private::FileSystem::Initialize<>() + 9 at FileSystem.h:51:5, address = 0x00007fffed8c6bf0, resolved, hit count = 1 
  1.3: where = liblldb.so.20.0git`lldb_private::SystemInitializerCommon::Initialize() + 252 [inlined] void lldb_private::FileSystem::Initialize<>() at FileSystem.h:51:5, address = 0x00007fffed8c6ccc, resolved, hit count = 0 
```

I think I've ended up with a second instance of FileSystem.h by referencing it in lldb-dap. I don't see errors like this on macOS, but that may be due to linker differences. I was using the `FileSystem().Fopen()` call to open `/dev/null` for the SBDebugger input handle.

I know in my other pull request we talked about linking against lldb/host for some utilities like the use of Pipe, but is this an issue with the current architecture? Should I avoid that Host/FileSystem API or will I run into this with other APIs potentially?

ashgti committed:
Fixing linux tests.

ashgti commented:
I have the linux tests working by avoiding the `Host/FileSystem` API, let me know if I should adjust the linkage though.

labath reviewed (approved):
> I know in my other pull request we talked about linking against lldb/host for some utilities like the use of Pipe, but is this an issue with the current architecture? Should I avoid that Host/FileSystem API or will I run into this with other APIs potentially?

You're right. This is potential problem, one whose scope I did not fully realize when I made that suggestion. It happens because liblldb only exposes the public lldb API, which means that anything linking against it gets its own copy it. This can be a problem with classes that use singletons or some for of explicit initialization, etc.

Nonetheless, I still believe reimplementing any class that we might want to use on the other side of the API is not a solution. Overall, I think you've managed to strike a good balance with this version of the patch. The filesystem class is not particularly important here as the only thing we really get from it is the (os-specific) name of /dev/null. The socket and pipe classes on the other hand, are a lot more complicated, which makes it more worthwhile to reuse them. Fortunately, they are also the ones that don't suffer from these kinds of problems.

If you run into other problems with some changes like this, let me know, and I'll see what I can do about it.

ashgti committed:
Update lldb/tools/lldb-dap/DAP.cpp

Co-authored-by: Pavel Labath <pavel@labath.sk>

ashgti commented:
@walter-erquinigo Wanted to double check with you on this since this change now will always link `lldb-dap` with `lldbHost`

walter-erquinigo commented:
@ashgti , this looks very nice! It's good overall that we are able to reuse some components and reduce the number of responsibilities in lldb-dap.

luporl commented:
> LLVM Buildbot has detected a new failure on builder `lldb-aarch64-windows` running on `linaro-armv8-windows-msvc-05` while building `lldb` at step 6 "test".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/141/builds/5258
> Here is the relevant piece of the build log for the reference

Hi. `lldb-aarch64-windows` has been broken for some time. Are there any plans to fix or revert this PR?

d0k commented:
Reverted this in 81898ac00e04ed3f352534a810829bdf4e6e14b7. Apart from breaking windows, I'm also seeing some weirdness when debugging iOS apps with this change, don't know what's going on.

ashgti commented:
I'll take a look
